[export] allow user input mutation in aot_export

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119356

This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

suo committed:
[export] allow user input mutation in aot_export

This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

[ghstack-poisoned]

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -439,6 +356,11 @@ def _export_non_strict(
             trace_joint=False,
             pre_dispatch=pre_dispatch,
         )
+    # TODO unfortunately preserving graph-level metadata is not
```
🤔 in theory the stuff Sherlock added for preserving nn stack traces from dynamo also means that node meta should be preserved

bdhirsh replied:
Oh, graph level meta

bdhirsh reviewed (approved):
Lgtm!

suo committed:
Update on "[export] allow user input mutation in aot_export"


This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

[ghstack-poisoned]

zhxchen17 reviewed (changes requested):
change looks good, but I prefer the previous node name convention which was regressed...

zhxchen17 commented on test/export/test_export.py:
```diff
@@ -1207,14 +1207,14 @@ def forward(self, x):
             """\
 def forward(self, arg_0):
-    l_x_, = fx_pytree.tree_flatten_spec(([arg_0], {}), self._in_spec)
+    arg7_1, = fx_pytree.tree_flatten_spec(([arg_0], {}), self._in_spec)
```
I think for debugging purposes I'd like to keep the original node names like `l_x_`

zhxchen17 commented on torch/export/_trace.py:
```diff
@@ -329,88 +328,6 @@ def _export_to_torch_ir(
-    return user_input_names
-
-
-def _lift_buffers_to_user_inputs(
```
but removing this is kinda nice.

zhxchen17 reviewed (approved):
after offline discussion, let's improve node names as a follow up after this diff

suo committed:
Update on "[export] allow user input mutation in aot_export"


This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

[ghstack-poisoned]

suo committed:
Update on "[export] allow user input mutation in aot_export"


This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

[ghstack-poisoned]

thiagocrepaldi commented:
@titaiwangms FYI

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -799,9 +799,9 @@ def model_signature(self) -> Optional[torch.export.ExportGraphSignature]:
                     InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='arg5_1'),
                               target='my_buffer1', persistent=True),
-                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='l_x_'),
+                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='arg6_1'),
```
Isnt it possible to preserve the name instead?

suo replied:
trickier with the current approach. It also was not guaranteed that these names would look the way they do if strict=False. I plan to followup with a more comprehensive input processing pass that should name these more nicely.

thiagocrepaldi replied:
> trickier with the current approach. It also was not guaranteed that these names would look the way they do if strict=False. I plan to followup with a more comprehensive input processing pass that should name these more nicely.

is there an issue filed for this? would love to know more about it

suo replied:
No, was just the conversation on this PR. Feel free to file an issue which we will keep updated

thiagocrepaldi replied:
https://github.com/pytorch/pytorch/issues/119484 thanks

thiagocrepaldi commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -1629,32 +1617,32 @@ def skip_torchlib_forward_compatibility(
         reason="does not match node name",
     ),
-    xfail(
+    skip(
```
why changing xfail by skip?

suo replied:
could use some help investigating this; for some reason if I left `xfail` then I was getting "unexpected success", but if I removed the xfail I was getting a failure…

thiagocrepaldi replied:
@liqunfu @titaiwangms @BowenBao was this the behavior you noticed this week too? what was the fix?

thiagocrepaldi replied:
> could use some help investigating this; for some reason if I left `xfail` then I was getting "unexpected success", but if I removed the xfail I was getting a failure…

@suo we will get back on this for you

titaiwangms replied:
Put it down to the list SKIP_XFAIL_SUBTESTS with xfail should work. Ithink I met this kind of weird op tests before.

suo commented:
@pytorchbot merge -f "landed internally"
